python/src/scipp/plot/tools.py

```diff
@@ -117,21 +117,56 @@ def vars_to_err(v):
     return v
 
 
-def check_log_limits(lims=None, vmin=None, vmax=None, scale=None):
-    """
-    Check if a limit is negative when a "log" norm is used.
-    If so, set the lower limits as 1.0e-3 * max_value.
-    Input can either be:
-      - 2 values `vmin` and `vmax` -> return vmin, vmax
-      - a list of length 2 `lims` -> return [vmin, vmax]
-    """
-    if lims is not None:
-        vmin = lims[0]
-        vmax = lims[1]
-    if vmin is not None and vmax is not None:
-        if scale == "log" and vmin <= 0:
-            vmin = 1.0e-03 * vmax
-    if lims is not None:
-        return np.array([vmin, vmax])
+def find_log_limits(x):
+    """
+    To find log scale limits, we histogram the data between 1.0-30
+    and 1.0e+30 and include only bins that are non-zero.
+    """
+    volume = np.product(x.shape)
+    pixel = sc.reshape(x, dims=['pixel'], shape=(volume, ))
+    pixel.variances = None  # no variances allowed in coord for histogram
+    weights = sc.Variable(dims=['pixel'], values=np.ones(volume))
+    hist = sc.histogram(sc.DataArray(data=weights, coords={'order': pixel}),
+                        bins=sc.Variable(dims=['order'],
+                                         values=np.geomspace(1e-30,
+                                                             1e30,
+                                                             num=61),
+                                         unit=x.unit))
+    # Find the first and the last non-zero bins
+    has_data = hist.data > 0.0 * sc.units.one
+    inds = np.where(has_data.values)
+    ar = np.arange(hist.data.shape[0])[inds]
+    # Safety check in case there are no positive values:
+    # return arbitrary values that ensure toggling log on the plot with buttons
+    # does not fail.
+    if len(ar) == 0:
+        vmin = 0.1
+        vmax = 1.0
```

Comment:
Should we fallback to `find_linear_limits` instead? Or something slow giving the right limits? Or the old mechanism?

Reply:
This is the case where there are no positive values in the data, so there are no "right" limits for a log scale.
I tried setting both to `None` but this was causing errors elsewhere (they seemed to then be overwritten somewhere else with negative values).
I can try and debug further (maybe it's better to understand fully what is happening), or keep it like this for now, which is has a stable behaviour.

Reply:
actually it's not being overwritten, it's a matplotlib behaviour: if you supply `vmin=None` to a `LogNorm()` for a figure where there are only negative values, it complains that `minvalue must be positive`.

Reply:
What about data outside the `1e+-30` limits?

Reply:
should we just extend the range above then?
you mean if someone is plotting stuff which has values between 1.0e-50 and 1.0e-30?

Reply:
> you mean if someone is plotting stuff which has values between 1.0e-50 and 1.0e-30?

Yes.

But by how much would you extend it? I am not sure extending can be the ultimate solution?

Reply:
so then find a min and a max manually, with `sc.min` and `sc.max`, then if both are negative just return maybe 0.1 and 1.0, and if max is positive, then set min to 1.0e-3 * max?

Reply:
Yes, something like that 👍